### PR TITLE
Convert TagEditorParam-related logic to TypeScript and add tests

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/29881

Changes:
- Adds fingerprints to `sample_database` mock preset
- Moves `TemplateTag` types to `metabase-types/api`
- Converts `tag-template-options` to TS 
- Converts tests for `TagEditorParam` to TS

How to verify:
- `yarn type-check` and `yarn test-unit` passes

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29411)

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/template_tags/TagEditorParam.unit.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        type: "number",
+        default: undefined,
+        dimension: undefined,
+        "widget-type": undefined,
+      });
+    });
+  });
+
+  describe("tag dimension", () => {
+    it("should default to string/contains for a new field filter", async () => {
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/template_tags/TagEditorParam.unit.spec.tsx:110]: 2 new tests here

File: frontend/src/metabase/query_builder/components/template_tags/TagEditorParam.jsx [hunk 2]
```diff
@@ -120,20 +123,10 @@ export class TagEditorParam extends Component {
...
-      ) {
-        widgetType = tag["widget-type"];
-      } else if (options.length > 0) {
-        widgetType = options[0].type;
-      }
       setTemplateTag({
         ...tag,
         dimension,
-        "widget-type": widgetType,
+        "widget-type": getDefaultParameterWidgetType(tag, field),
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/components/template_tags/TagEditorParam.jsx:129]: If everything works out I'll migrate `TagEditorParam` to typescript in the next PR

**Comment by mazameli** [comment]
This doesn't seem to be defaulting to `Contains` anymore for me, @ranquild. Do I need to be selecting a field with higher cardinality now?

image

File: e2e/test/scenarios/admin/datamodel/metrics.cy.spec.js [hunk 3]
```diff
@@ -166,7 +166,9 @@ describe("scenarios > admin > datamodel > metrics", () => {
       cy.visit("/reference/metrics/1/questions");
       cy.findByText("Our analysis").should("not.exist");
       cy.findAllByText("Questions about orders < 100");
-      cy.findByText("Orders, orders < 100, Filtered by Total is greater than or equal to 50");
+      cy.findByText(
```
Comment (ranquild) [inline e2e/test/scenarios/admin/datamodel/metrics.cy.spec.js:169]: Prettier changed this for some reason

File: frontend/src/metabase-lib/parameters/utils/template-tag-options.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import Field from "metabase-lib/metadata/Field";
+import {
+  ID_OPTION,
+  OPTIONS_WITH_OPERATOR_SUBTYPES,
+  PARAMETER_OPERATOR_TYPES,
+} from "../constants";
+import { buildTypedOperatorOptions, getOperatorDisplayName } from "./operators";
+import { fieldFilterForParameter } from "./filters";
+
+export function getParameterOptions() {
```
Comment (ranquild) [inline frontend/src/metabase-lib/parameters/utils/template-tag-options.ts:12]: JS -> TS

File: frontend/src/metabase-lib/parameters/utils/template-tag-options.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    .filter(option => fieldFilterForParameter(option)(field))
+    .map(option => {
+      return {
+        ...option,
+        name: "combinedName" in option ? option.combinedName : option.name,
+      };
+    });
+}
+
+export function getDefaultParameterWidgetType(tag: TemplateTag, field: Field) {
```
Comment (ranquild) [inline frontend/src/metabase-lib/parameters/utils/template-tag-options.ts:36]: The only new function here - extracted from `TagEditorParam` component

File: frontend/src/metabase-types/api/field.ts [hunk 6]
```diff
@@ -21,22 +37,10 @@ export type NumberFieldFingerprint = {
 };
 
 export type DateTimeFieldFingerprint = {
-  earliest: "2016-04-26T19:29:55.147Z";
-  latest: "2019-04-15T13:34:19.931Z";
+  earliest: string;
```
Comment (oisincoveney) [inline frontend/src/metabase-types/api/field.ts:40]: Would it make sense to have a more specific type here? Obviously using `string` works but I was thinking that we could use something that could catch date strings with typescript so we don't run into something weird with badly-formatted date strings like the one specified [here](https://gist.github.com/MrChocolatine/367fb2a35d02f6175cc8ccb3d3a20054) - obviously it's a little more code but might help in the future?

We could then union the ISO string type with the `YY-mm-dd` to accept the data in `createMockDateTimeFieldFingerprint`

File: frontend/src/metabase-lib/parameters/utils/template-tag-options.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  const options = getParameterOptionsForField(field);
+  const widgetType = tag["widget-type"];
+
+  if (options.length === 0) {
```
Comment (uladzimirdev) [inline frontend/src/metabase-lib/parameters/utils/template-tag-options.ts:40]: nit: 

if would simplify it a bit:

```
if (options.length === 0) {
  return
}

if (widgetType ...) {

} else {

}
```

**Review by uladzimirdev — APPROVED** [review]
LGTM overall, left a few comments

File: frontend/src/metabase-lib/parameters/utils/template-tag-options.ts (hunk 7 above)
Comment (ranquild) [reply frontend/src/metabase-lib/parameters/utils/template-tag-options.ts:38]: Fixed, thanks!

File: frontend/src/metabase/query_builder/components/template_tags/TagEditorParam.jsx [hunk 8]
```diff
@@ -274,17 +267,17 @@ export class TagEditorParam extends Component {
 
         {(hasWidgetOptions || !isDimension) && (
           <InputContainer>
-            <ContainerLabel>
+            <ContainerLabel htmlFor="tag-editor-display-name">
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase/query_builder/components/template_tags/TagEditorParam.jsx]: Made `InputContainer` be a `label` and removed `htmlFor`. Thanks!

**Pull request merged**